File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function (head) {
    var slow = head;
    var fast = head;

    var hascycle = false;
    while (fast != null && fast.next != null) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow == fast) {
            hascycle = true;
            break;

        }

    }
    console.log({ hascycle })
    if (hascycle) {
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;


        }
        if (slow == fast) {
            return slow;
        }
    }
    return null
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* detectCycle(ListNode* head) {
        bool iscylic = false;
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                iscylic = true;
                break;
            }
        }
        if (!iscylic)
            return NULL;
        slow = head;

        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};